### ISSUE TITLE
Unnecessary cast to float32 in ArrayField

### ISSUE THREAD

**Issue by festeh** [allenai/allennlp#2454]
**Describe the bug**
    ArrayField silently casts the array to float32 during as_tensor call

**To Reproduce**
```
array = np.array([1, 2, 3])
field = ArrayField(array)
tensor = field.as_tensor(field.get_padding_lengths())
print(array.dtype, tensor.dtype)
```

**Expected behavior**
    I would expect that padding operation generally should not change array type. I suppose that `self.array.dtype` information should be used, if available.

**Comment by matt-gardner** [comment]
Looks like the problem might be this line: https://github.com/allenai/allennlp/blob/dddcbf243e9931a5ab0241435b39613117e20acf/allennlp/data/fields/array_field.py#L31

PRs welcome!

**Comment by DeNeutoy** [comment]
Closed by #2609
